src/trace/harTracer.ts

```diff
@@ -0,0 +1,311 @@
+import * as network from '../server/network';
+import { Page } from '../server/page';
+import * as har from './har';
+
+const fsWriteFileAsync = util.promisify(fs.writeFile.bind(fs));
+
+export function installHarTracer() {
+  contextListeners.add(new HarTracer());
+}
+
+class HarTracer implements ContextListener {
+  private _contextTracers = new Map<BrowserContext, HarContextTracer>();
+
+  async onContextCreated(context: BrowserContext): Promise<void> {
+    if (!context._options.recordHar)
+      return;
+    const contextTracer = new HarContextTracer(context, context._options.recordHar);
+    this._contextTracers.set(context, contextTracer);
+  }
+
+  async onContextWillDestroy(context: BrowserContext): Promise<void> {
+    const contextTracer = this._contextTracers.get(context);
+    if (contextTracer) {
+      this._contextTracers.delete(context);
+      await contextTracer.flush();
+    }
+  }
+
+  async onContextDidDestroy(context: BrowserContext): Promise<void> {
+    const contextTracer = this._contextTracers.get(context);
+    if (contextTracer) {
+      this._contextTracers.delete(context);
+      await contextTracer.flush();
+    }
+  }
+}
+
+type HarOptions = {
+  path: string;
+  omitContent?: boolean;
+};
+
+class HarContextTracer {
+  private _options: HarOptions;
+  private _browserName: string;
+  private _log: har.Log;
+  private _pageEntries = new Map<Page, har.Page>();
+  private _entries = new Map<network.Request, har.Entry>();
+  private _lastPage = 0;
+  private _barrierPromises = new Map<Promise<void>, Page>();
+
+  constructor(context: BrowserContext, options: HarOptions) {
+    this._browserName = context._browser._options.name;
+    this._options = options;
+    this._log = {
+      version: '1.2',
+      creator: {
+        name: 'Playwright',
+        version: require('../../package.json')['version'],
+      },
+      browser: {
+        name: context._browser._options.name,
+        version: context._browser.version()
+      },
+      pages: [],
+      entries: []
+    };
+    context.on(BrowserContext.Events.Page, page => this._onPage(page));
+  }
+
+  private _onPage(page: Page) {
+    const pageEntry: har.Page = {
+      startedDateTime: new Date(),
+      id: `page_${this._lastPage++}`,
+      title: '',
+      pageTimings: {
+        onContentLoad: -1,
+        onLoad: -1,
+      },
+    };
+    this._pageEntries.set(page, pageEntry);
+    this._log.pages.push(pageEntry);
+    page.on(Page.Events.Request, (request: network.Request) => this._onRequest(page, request));
+    page.on(Page.Events.Response, (response: network.Response) => this._onResponse(page, response));
+
+    page.on(Page.Events.DOMContentLoaded, () => {
+      const promise = page.mainFrame()._evaluateExpression(String(() => {
+        return {
+          title: document.title,
+          domContentLoaded: performance.timing.domContentLoadedEventStart,
+        };
+      }), true, undefined).then(result => {
```

Comment:
```suggestion
      }), true, 'utility').then(result => {
```
People will cheat on their lighthouse scores otherwise ;)

Reply:
Wrong param, but sure.

Reply:
Done.